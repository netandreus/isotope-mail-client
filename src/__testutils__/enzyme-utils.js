import {shallowToJson} from 'enzyme-to-json';

/**
 * Serialized the component using enzyme-to-json removing unnecessary and ugly store properties
 * @param component
 * @returns {*}
 */
export const shallowComponentWithStoreToJson = component =>
  shallowToJson(component, {map: json => {
    const props = {...json.props};
    delete props.dispatch;
    delete props.store;
    delete props.storeSubscription;
    return ({...json, props: props});
  }});
