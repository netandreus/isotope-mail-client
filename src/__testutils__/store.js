import {createStore} from 'redux';
import rootReducer, {INITIAL_STATE} from '../reducers';

export const createMockStore = initialState => createStore(rootReducer, initialState);
export const MOCK_STORE = createMockStore(INITIAL_STATE);
