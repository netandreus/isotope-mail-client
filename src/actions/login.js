import {ActionTypes} from './action-types';

export const setFormValues = formValues => ({
  type: ActionTypes.LOGIN_FORM_VALUES_SET, payload: formValues
});
