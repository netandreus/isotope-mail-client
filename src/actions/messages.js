import {ActionTypes} from './action-types';

export const backendRequest = () => ({type: ActionTypes.MESSAGES_BE_REQUEST});
export const backendRequestCompleted = () => ({type: ActionTypes.MESSAGES_BE_REQUEST_COMPLETED});

export const setCache = cache => ({
  type: ActionTypes.MESSAGES_SET_CACHE,
  payload: cache
});

export const setFolderCache = (folder, messages) => ({
  type: ActionTypes.MESSAGES_SET_FOLDER_CACHE,
  payload: {folder, messages}
});

/**
 * Updates state folder's messages cache.
 *
 * @param folder to which the messages will be added
 * @param messages {Array} array of messages to add to the folder's cache
 * @returns {{type: string, payload: {folder: *, messages: *}}}
 */
export const updateCache = (folder, messages) => ({
  type: ActionTypes.MESSAGES_UPDATE_CACHE,
  payload: {folder, messages}
});

/**
 * Updates state folder's messages cache only if the messages already exist in the folder cache.
 *
 * @param folder to which the messages will be added
 * @param messages {Array} array of messages to add to the folder's cache
 * @param {boolean} ignoreLocked
 * @returns {{type: string, payload: {folder: *, messages: *}}}
 */
export const updateCacheIfExist = (folder, messages, ignoreLocked = false) => ({
  type: ActionTypes.MESSAGES_UPDATE_CACHE_IF_EXIST,
  payload: {folder, messages, ignoreLocked}
});

/**
 * Delete messages from state folder's messages cache.
 *
 * @param {!object} folder from which the messages will be deleted
 * @param {!Array} messages array of messages to delete from the folder's cache
 * @param {object} [deleteUidRange] Range of message UIDs to keep (inclusive)
 * @param {number} [deleteUidRange.from] delete all messages starting with this UID
 * @param {number} [deleteUidRange.to] delete all messages until this UID
 * @returns {{type: string, payload: {folder: *, messages: *}}}
 */
export const deleteFromCache = (folder, messages, deleteUidRange) => ({
  type: ActionTypes.MESSAGES_DELETE_FROM_CACHE,
  payload: {folder, messages, deleteUidRange}
});

export const renameCache = (oldId, newId) => ({
  type: ActionTypes.MESSAGES_RENAME_CACHE,
  payload: {oldId, newId}
});

/**
 * Adds the provided message uid to the selected message list if selected or else removes it from the list
 *
 * @param message {Object}
 * @param selected {boolean}
 * @returns {{type: string, payload: {message: *, selected: *}}}
 */
export const setSelected = (messages, selected) => ({
  type: ActionTypes.MESSAGES_SET_SELECTED,
  payload: {messages, selected}
});

/**
 * Clears the list of selected message uids
 *
 * @returns {{type: string}}
 */
export const clearSelected = () => ({
  type: ActionTypes.MESSAGES_CLEAR_SELECTED
});

/**
 * Adds the messageId(s) of the provided messages to the messages.locked array in the Redux store.
 *
 * @param {Array.<{messageId: string}>} messages
 * @returns {{type: string, payload: *}}
 */
export const lockMessages = messages => ({
  type: ActionTypes.MESSAGES_LOCK_ADD, payload: messages
});

/**
 * Removes the messageId(s) of the provided messages from the messages.locked array in the Redux store.
 *
 * @param {Array.<{messageId: string}>} messages
 * @returns {{type: string, payload: *}}
 */
export const unlockMessages = messages => ({
  type: ActionTypes.MESSAGES_LOCK_REMOVE, payload: messages
});
