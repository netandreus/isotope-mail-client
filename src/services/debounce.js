/**
 * Debounce a function so that it get called only once in (at minimum) the specified period
 *
 * @param func to be debounced
 * @param period in milliseconds
 * @returns {Function} The debounced function
 */
function debounce(func, period = 100) {
  let funcTimeout;
  function debounced() {
    const funcContext = this;
    const funcArguments = arguments;
    const runDebounced = () => {
      funcTimeout = null;
      func.apply(funcContext, funcArguments);
    };
    clearTimeout(funcTimeout);
    funcTimeout = setTimeout(runDebounced, period);
  }
  debounced.cancel = function () {
    clearTimeout(funcTimeout);
  };
  return debounced;
}


export default debounce;
