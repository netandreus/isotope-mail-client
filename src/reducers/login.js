import {INITIAL_STATE} from './index';
import {ActionTypes} from '../actions/action-types';


const login = (state = INITIAL_STATE.login, action = {}) => {
  if (action.type === ActionTypes.LOGIN_FORM_VALUES_SET) {
    const newState = {...state, formValues: {...action.payload}};
    delete newState.formValues.password;
    return newState;
  }
  return state;
};

export default login;
