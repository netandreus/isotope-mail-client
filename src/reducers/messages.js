import {INITIAL_STATE} from './';
import {ActionTypes} from '../actions/action-types';

const messages = (state = INITIAL_STATE.messages, action = {}) => {
  switch (action.type) {
    case ActionTypes.MESSAGES_BE_REQUEST:
      return {...state, activeRequests: state.activeRequests + 1};
    case ActionTypes.MESSAGES_BE_REQUEST_COMPLETED:
      return {...state, activeRequests: state.activeRequests > 0 ? state.activeRequests - 1 : 0};
    case ActionTypes.MESSAGES_SET_CACHE:
      return {...state, cache: action.payload};
    case ActionTypes.MESSAGES_SET_FOLDER_CACHE: {
      const newState = {...state};
      newState.cache = {...newState.cache};
      newState.cache[action.payload.folder.folderId] = new Map(action.payload.messages.map(m => [m.uid, m]));
      return newState;
    }
    case ActionTypes.MESSAGES_UPDATE_CACHE: {
      const newUpdateState = {...state};
      newUpdateState.cache = {...newUpdateState.cache};
      if (newUpdateState.cache[action.payload.folder.folderId] instanceof Map === false) {
        newUpdateState.cache[action.payload.folder.folderId] = new Map();
      } else {
        newUpdateState.cache[action.payload.folder.folderId] =
          new Map(newUpdateState.cache[action.payload.folder.folderId]);
      }
      action.payload.messages
        .filter(m => !m.messageId || !newUpdateState.locked.includes(m.messageId))
        .forEach(m => newUpdateState.cache[action.payload.folder.folderId].set(m.uid, m));
      return newUpdateState;
    }
    case ActionTypes.MESSAGES_UPDATE_CACHE_IF_EXIST: {
      const newUpdateState = {...state};
      newUpdateState.cache = {...newUpdateState.cache};
      if (newUpdateState.cache[action.payload.folder.folderId] instanceof Map === true) {
        newUpdateState.cache[action.payload.folder.folderId] =
          new Map(newUpdateState.cache[action.payload.folder.folderId]);
        const cache = newUpdateState.cache[action.payload.folder.folderId];
        let messagesToUpdate;
        if (action.payload.ignoreLocked) {
          messagesToUpdate = action.payload.messages;
        } else {
          messagesToUpdate = action.payload.messages
            .filter(m => !m.messageId || !newUpdateState.locked.includes(m.messageId));
        }
        messagesToUpdate.forEach(m => {
          if (cache.has(m.uid)) {
            cache.set(m.uid, m);
          }
        });
      }
      return newUpdateState;
    }
    case ActionTypes.MESSAGES_DELETE_FROM_CACHE: {
      const newUpdateState = {...state};
      if (newUpdateState.cache[action.payload.folder.folderId] instanceof Map === false) {
        return state;
      }

      newUpdateState.cache = {...newUpdateState.cache};
      newUpdateState.cache[action.payload.folder.folderId] =
        new Map(newUpdateState.cache[action.payload.folder.folderId]);
      // Delete unitary messages
      const cache = newUpdateState.cache[action.payload.folder.folderId];
      action.payload.messages.forEach(m => cache.delete(m.uid));

      // Delete UID range
      const {deleteUidRange} = action.payload;
      if (deleteUidRange) {
        let toDeleteUids = Array.from(cache.keys());
        if (deleteUidRange.from || deleteUidRange.from === 0) {
          toDeleteUids = toDeleteUids.filter(uid => uid >= deleteUidRange.from);
        }
        if (deleteUidRange.to || deleteUidRange.to === 0) {
          toDeleteUids = toDeleteUids.filter(uid => uid <= deleteUidRange.to);
        }
        toDeleteUids.forEach(uid => cache.delete(uid));
      }
      return newUpdateState;
    }
    case ActionTypes.MESSAGES_RENAME_CACHE: {
      const newState = {...state};
      if (newState.cache[action.payload.oldId]) {
        newState.cache = {...newState.cache};
        newState.cache[action.payload.newId] = newState.cache[action.payload.oldId];
        delete newState.cache[action.payload.oldId];
      }
      return newState;
    }
    case ActionTypes.MESSAGES_SET_SELECTED: {
      const newUpdateState = {...state};
      newUpdateState.selected = [...state.selected];
      action.payload.messages.forEach(message => {
        const indexOfMessage = newUpdateState.selected.indexOf(message.uid);
        if (action.payload.selected && indexOfMessage < 0) {
          // Select Message
          newUpdateState.selected.push(message.uid);
        } else if (!action.payload.selected) {
          // Unselect message
          newUpdateState.selected = newUpdateState.selected.filter(uid => uid !== message.uid);
        }
      });
      return newUpdateState;
    }
    case ActionTypes.MESSAGES_CLEAR_SELECTED: {
      const newUpdateState = {...state, selected: []};
      return newUpdateState;
    }
    case ActionTypes.MESSAGES_LOCK_ADD: {
      const newState = {...state};
      newState.locked = [
        ...newState.locked,
        ...action.payload
          .filter(m => m.messageId && m.messageId.length > 0)
          .map(m => m.messageId)
      ];
      return newState;
    }
    case ActionTypes.MESSAGES_LOCK_REMOVE: {
      const newState = {...state};
      const remainingItems = [
        ...action.payload.filter(m => m.messageId && m.messageId.length > 0).map(m => m.messageId)
      ];
      newState.locked = [];
      for (const messageId of state.locked) {
        const i = remainingItems.indexOf(messageId);
        if (i > -1) {
          remainingItems.splice(i, 1);
        } else {
          newState.locked.push(messageId);
        }
      }
      return newState;
    }
    default:
      return state;
  }
};

export default messages;
