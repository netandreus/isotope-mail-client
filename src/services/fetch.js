import {refreshUserCredentials} from '../actions/application';

export const HttpHeaders = {
  ISOTOPE_CREDENTIALS: 'Isotope-Credentials',
  ISOTOPE_SALT: 'Isotope-Salt',
  CONTENT_TYPE: 'Content-Type'
};

export const HttpStatusFamilies = {
  INFORMATIONAL: 1,
  SUCCESSFUL: 2,
  REDIRECTION: 3,
  CLIENT_ERROR: 4,
  SERVER_ERROR: 5
};

export class AuthenticationException extends Error {
  constructor(message, cause) {
    super(message);
    this.cause = cause;
    this.name = 'AuthenticationException';
  }
}

/**
 * Object to store the different AbortController(s) that will be used in the service methods to fetch from the API backend.
 *
 * @type {Object}
 * @private
 */
export const abortControllerWrappers = {};

/**
 * Converts response Promise to json Promise or throws an error.
 *
 * @param {Response}response
 * @returns {PromiseLike<object>}
 */
export function toJson(response) {
  if (!response.ok) {
    if (response.status === 401 || response.status === 403) {
      throw new AuthenticationException(response.statusText);
    } else {
      throw Error(`${response.status}`);
    }
  }
  return response.json();
}

/**
 * Dispatches the refreshUserCredentials action with the values obtained from the Http Response headers.
 *
 * @param dispatch
 * @param response
 * @returns {*}
 */
export function refreshCredentials(dispatch, response) {
  const credentials = response.headers.get(HttpHeaders.ISOTOPE_CREDENTIALS);
  const salt = response.headers.get(HttpHeaders.ISOTOPE_SALT);
  if (credentials && salt) {
    dispatch(refreshUserCredentials(credentials, salt));
  }
  return Promise.resolve(response);
}

/**
 * Triggers the abort method of a the specified AbortController
 *
 * @param abortController {AbortController}
 */
export function abortFetch(abortController) {
  if (abortController && abortController.abort) {
    abortController.abort();
  }
}

/**
 * Creates a new headers object including Isotope specific credentials headers for server authentication.
 *
 * @param {{}} credentials
 * @param {{}} [originalHeaders={}] originalHeaders
 * @returns {{}}
 */
export function credentialsHeaders(credentials, originalHeaders) {
  const newHeaders = Object.assign({}, originalHeaders ? originalHeaders : {});
  newHeaders[HttpHeaders.ISOTOPE_CREDENTIALS] = credentials.encrypted;
  newHeaders[HttpHeaders.ISOTOPE_SALT] = credentials.salt;
  return newHeaders;
}

/**
 * Returns true if the response the Successful HTTP status code range (2xx)
 *
 * @param HTTP status code of the response
 * @returns {boolean} true if status code belongs to HttpStatusFamilies.SUCCESSFUL family
 */
export function isSuccessful(status) {
  return Math.floor(status / 100) === HttpStatusFamilies.SUCCESSFUL;
}

