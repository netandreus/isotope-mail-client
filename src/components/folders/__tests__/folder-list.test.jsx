import React from 'react';
import {shallow} from 'enzyme';
import {createMockStore, MOCK_STORE} from '../../../__testutils__/store';
import {INITIAL_STATE} from '../../../reducers';
import FolderList, {FolderListClass} from '../folder-list';
import * as folderService from '../../../services/folder';
import * as messageService from '../../../services/message';
import {ActionTypes} from '../../../actions/action-types';
import {FolderTypes} from '../../../services/folder';

describe('FolderList component test suite', () => {
  describe('Snapshot render', () => {
    test('Inbox; folder and folder with children, Should render FolderListClass', () => {
      const selectedFolder = {
        folderId: 1, type: FolderTypes.INBOX, name: 'Inbox',
        unreadMessageCount: 1, newMessageCount: 1, children: []};
      // Given
      const props = {
        folderTree: [
          selectedFolder,
          {folderId: 2, children: []},
          {folderId: 3, children: [{folderId: 4, children: []}]}
        ],
        folders: {
          1: selectedFolder,
          2: {
            folderId: 2, type: FolderTypes.FOLDER, name: 'Folder 1',
            unreadMessageCount: 1, newMessageCount: 1, children: []
          },
          3: {
            folderId: 3, type: FolderTypes.FOLDER, name: 'Folder 2',
            unreadMessageCount: 1, newMessageCount: 1, children: [{folderId: 4, children: []}
            ]
          },
          4: {
            folderId: 4, type: FolderTypes.FOLDER, name: 'Folder 1',
            unreadMessageCount: 1, newMessageCount: 1, children: []
          }
        },
        selectedFolder: selectedFolder
      };

      // When
      const folderList = shallow(<FolderListClass {...props} />);

      // Then
      expect(folderList).toMatchSnapshot();
    });
  });
  test('selectFolder, dispatch actions triggered', () => {
    // Given
    let dispatchCount = 0;
    const store = createMockStore(INITIAL_STATE);
    store.dispatch = jest.fn(action => {
      switch (action.type) {
        case ActionTypes.APPLICATION_FOLDER_SELECT:
        case ActionTypes.APPLICATION_MESSAGE_SELECT:
        case ActionTypes.MESSAGES_CLEAR_SELECTED:
          dispatchCount++;
          break;
        default:
      }
    });
    messageService.resetFolderMessagesCache = jest.fn();
    const props = {
      activeRequests: 0, folderList: [{folderList: []}], selectedFolder: {folderId: '1337-3'}};
    const folderList = shallow(<FolderList store={store} {...props}/>);

    // When
    folderList.props().selectFolder({});

    // Then
    expect(dispatchCount).toEqual(3);
    expect(messageService.resetFolderMessagesCache).toHaveBeenCalledTimes(1);
  });
  test('renameFolder, dispatch actions triggered', done => {
    // Given
    const store = createMockStore(INITIAL_STATE);
    store.dispatch = jest.fn(action => {
      if (action.type === ActionTypes.APPLICATION_FOLDER_RENAME) {
        expect(action.payload.folderId).toEqual('1337');
        done();
      }
    });
    const props = {
      activeRequests: 0, folderList: [{folderList: []}], selectedFolder: {folderId: '1337-3'}};
    const folderList = shallow(<FolderList store={store} {...props}/>);

    // When
    folderList.props().renameFolder({folderId: '1337'});

    // Then
    // Async assertions in APPLICATION_FOLDER_RENAME action
  });
  test('createChildInFolder, dispatch actions triggered', done => {
    // Given
    const store = createMockStore(INITIAL_STATE);
    store.dispatch = jest.fn(action => {
      if (action.type === ActionTypes.APPLICATION_FOLDER_CREATE) {
        expect(action.payload).toEqual('1337');
        done();
      }
    });
    const props = {
      activeRequests: 0, folderList: [{folderList: []}], selectedFolder: {folderId: '1337-3'}};
    const folderList = shallow(<FolderList store={store} {...props}/>);

    // When
    folderList.props().createChildInFolder({folderId: '1337'});

    // Then
    // Async assertions in APPLICATION_FOLDER_CREATE action
  });
  test('deleteFolder, dispatch actions triggered', () => {
    // Given
    folderService.deleteFolder = jest.fn();
    const props = {
      activeRequests: 0, folderList: [{folderList: []}], selectedFolder: {folderId: '1337-3'}};
    const folderList = shallow(<FolderList store={MOCK_STORE} {...props}/>);

    // When
    folderList.props().deleteFolder({folderId: 'folderToDelete'});

    // Then
    expect(folderService.deleteFolder).toHaveBeenCalledTimes(1);
  });
  test('moveFolder, dispatch actions triggered', () => {
    // Given
    folderService.moveFolder = jest.fn();
    const props = {
      activeRequests: 0, folderList: [{folderList: []}], selectedFolder: {folderId: '1337-3'}};
    const folderList = shallow(<FolderList store={MOCK_STORE} {...props}/>);

    // When
    folderList.props().moveFolder({folderId: 'folderToMove'}, {folderId: 'targetFolder'});

    // Then
    expect(folderService.moveFolder).toHaveBeenCalledTimes(1);
  });
  test('moveMessages, service called', () => {
    // Given
    messageService.moveMessages = jest.fn();
    const props = {
      activeRequests: 0, folderList: [{folderList: []}], selectedFolder: {folderId: '1337-3'}};
    const folderContainer = shallow(<FolderList store={MOCK_STORE} {...props}/>);

    // When
    folderContainer.props().moveMessages();

    // Then
    expect(messageService.moveMessages).toHaveBeenCalledTimes(1);
  });
});
