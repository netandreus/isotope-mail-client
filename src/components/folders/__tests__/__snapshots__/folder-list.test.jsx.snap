// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FolderList component test suite Snapshot render Inbox; folder and folder with children, Should render FolderListClass 1`] = `
Array [
  <div
    className="itemContainer"
  >
    <Translate(FolderItem)
      className="item"
      draggable={false}
      graphic="inbox"
      label="Inbox"
      newMessageCount={1}
      onAddChild={[Function]}
      onClick={[Function]}
      onDelete={null}
      onDragStart={[Function]}
      onDrop={[Function]}
      onRename={null}
      selected={true}
      unreadMessageCount={1}
    />
  </div>,
  <div
    className="itemContainer"
  >
    <Translate(FolderItem)
      className="item"
      draggable={true}
      graphic="folder"
      label="Folder 1"
      newMessageCount={1}
      onAddChild={[Function]}
      onClick={[Function]}
      onDelete={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onRename={[Function]}
      selected={false}
      unreadMessageCount={1}
    />
  </div>,
  <div
    className="itemContainer"
  >
    <Translate(FolderItem)
      className="item"
      draggable={true}
      graphic="folder"
      label="Folder 2"
      newMessageCount={1}
      onAddChild={[Function]}
      onClick={[Function]}
      onDelete={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onRename={[Function]}
      selected={false}
      unreadMessageCount={1}
    />
    <nav
      className="mdc-list childList"
    >
      <Connect(FolderListClass)
        folderTree={
          Array [
            Object {
              "children": Array [],
              "folderId": 4,
            },
          ]
        }
        folders={
          Object {
            "1": Object {
              "children": Array [],
              "folderId": 1,
              "name": "Inbox",
              "newMessageCount": 1,
              "type": Object {
                "icon": "inbox",
                "position": 0,
                "serverName": "INBOX",
              },
              "unreadMessageCount": 1,
            },
            "2": Object {
              "children": Array [],
              "folderId": 2,
              "name": "Folder 1",
              "newMessageCount": 1,
              "type": Object {
                "icon": "folder",
              },
              "unreadMessageCount": 1,
            },
            "3": Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "folderId": 4,
                },
              ],
              "folderId": 3,
              "name": "Folder 2",
              "newMessageCount": 1,
              "type": Object {
                "icon": "folder",
              },
              "unreadMessageCount": 1,
            },
            "4": Object {
              "children": Array [],
              "folderId": 4,
              "name": "Folder 1",
              "newMessageCount": 1,
              "type": Object {
                "icon": "folder",
              },
              "unreadMessageCount": 1,
            },
          }
        }
      />
    </nav>
  </div>,
]
`;
